File: LeetCode/Java/Easy/PalindromeLinkedList.java
package Easy;

import java.util.*;

public class PalindromeLinkedList {

	public static void main(String[] args) {

	}
	
	public boolean isPalindrome(ListNode head){
		int len = 0;
		ListNode ptr = head;
		while (ptr != null){
			ptr = ptr.next;
			len++;
		}
		ptr = head;
		
		Stack<Integer> stk = new Stack();
		int count = 0;
		while(len/2 > count){
			stk.push(ptr.val);
			ptr = ptr.next;
			count++;
		}
		if (len %2 == 1) ptr = ptr.next;
		
		while (ptr != null){
			if (stk.pop() != ptr.val) return false;
			ptr = ptr.next;
		}
		return true;
	}
	


}

File: LeetCode/Python/Easy/PalindromeLinkedList.py
class Solution(object):
    def isPalindrome(self, head):
        fast, slow = head
        while (fast is not None and fast.next is not None):
            slow = slow.next
            fast = fast.next.next
        if (fast is not None):
            slow = slow.next
        slow = self.reverseList(slow)
        fast = head

        while (slow is not None):
            if(fast.val != slow.val):
                return False
            fast = fast.next
            slow = slow.next

    def reverseList(self, head):
        prev = None
        while (head is not None):
            next = head.next
            head.next = prev
            prev = head
            head = next
        return prev
